File: src/core/HexCoordinates.ts
import * as THREE from 'three';
import { HexMetrics } from './HexMetrics';
import { HexDirection, DirectionOffsets } from './HexDirection';

/**
 * Cube coordinates for hexagonal grids.
 * Uses the constraint: q + r + s = 0
 * This is an immutable value type.
 */
export class HexCoordinates {
  readonly q: number;  // Column-ish
  readonly r: number;  // Row-ish (diagonal)
  readonly s: number;  // Derived: -q - r

  constructor(q: number, r: number) {
    this.q = q;
    this.r = r;
    this.s = -q - r;
  }

  // Create from offset coordinates (rectangular grid)
  static fromOffset(col: number, row: number): HexCoordinates {
    // Odd-r offset coordinate system
    const q = col - Math.floor(row / 2);
    const r = row;
    return new HexCoordinates(q, r);
  }

  // Create from world position (XZ plane)
  static fromWorldPosition(position: THREE.Vector3): HexCoordinates {
    const { outerRadius, innerRadius } = HexMetrics;

    // Convert to fractional cube coordinates
    const q = (position.x * Math.sqrt(3) / 3 - position.z / 3) / outerRadius;
    const r = position.z * 2 / 3 / outerRadius;

    return HexCoordinates.round(q, r);
  }

  // Round fractional cube coordinates to nearest hex
  static round(q: number, r: number): HexCoordinates {
    const s = -q - r;

    let rq = Math.round(q);
    let rr = Math.round(r);
    let rs = Math.round(s);

    const qDiff = Math.abs(rq - q);
    const rDiff = Math.abs(rr - r);
    const sDiff = Math.abs(rs - s);

    // Fix rounding errors by recalculating the coordinate with largest diff
    if (qDiff > rDiff && qDiff > sDiff) {
      rq = -rr - rs;
    } else if (rDiff > sDiff) {
      rr = -rq - rs;
    }
    // s is always derived, no need to fix

    return new HexCoordinates(rq, rr);
  }

  // Convert to world position (center of hex)
  toWorldPosition(elevation: number = 0): THREE.Vector3 {
    const { outerRadius, innerRadius, elevationStep } = HexMetrics;

    const x = (this.q + this.r / 2) * innerRadius * 2;
    const z = this.r * outerRadius * 1.5;
    const y = elevation * elevationStep;

    return new THREE.Vector3(x, y, z);
  }

  // Get neighbor in specified direction
  getNeighbor(direction: HexDirection): HexCoordinates {
    const [dq, dr] = DirectionOffsets[direction];
    return new HexCoordinates(this.q + dq, this.r + dr);
  }

  // Get all 6 neighbors
  getNeighbors(): HexCoordinates[] {
    return [
      this.getNeighbor(HexDirection.NE),
      this.getNeighbor(HexDirection.E),
      this.getNeighbor(HexDirection.SE),
      this.getNeighbor(HexDirection.SW),
      this.getNeighbor(HexDirection.W),
      this.getNeighbor(HexDirection.NW),
    ];
  }

  // Distance to another hex (in hex steps)
  distanceTo(other: HexCoordinates): number {
    return Math.max(
      Math.abs(this.q - other.q),
      Math.abs(this.r - other.r),
      Math.abs(this.s - other.s)
    );
  }

  // Get all hexes in a ring at distance
  static ring(center: HexCoordinates, radius: number): HexCoordinates[] {
    if (radius === 0) return [center];

    const results: HexCoordinates[] = [];
    // Start at the hex radius steps in direction 4 (SW)
    let hex = new HexCoordinates(
      center.q + DirectionOffsets[HexDirection.SW][0] * radius,
      center.r + DirectionOffsets[HexDirection.SW][1] * radius
    );

    // Walk around the ring
    for (let dir = 0; dir < 6; dir++) {
      for (let i = 0; i < radius; i++) {
        results.push(hex);
        hex = hex.getNeighbor(dir as HexDirection);
      }
    }

    return results;
  }

  // Get all hexes in a spiral from center out to radius
  static spiral(center: HexCoordinates, radius: number): HexCoordinates[] {
    const results: HexCoordinates[] = [center];
    for (let r = 1; r <= radius; r++) {
      results.push(...HexCoordinates.ring(center, r));
    }
    return results;
  }

  // Equality check
  equals(other: HexCoordinates): boolean {
    return this.q === other.q && this.r === other.r;
  }

  // String key for use in Maps
  toKey(): string {
    return `${this.q},${this.r}`;
  }

  // Create from key string
  static fromKey(key: string): HexCoordinates {
    const [q, r] = key.split(',').map(Number);
    return new HexCoordinates(q, r);
  }

  toString(): string {
    return `Hex(${this.q}, ${this.r}, ${this.s})`;
  }
}

File: results/csharp_converted/organized/Core/hex_coordinates.cs
using Godot;
using Godot.Collections;


//# Axial hex coordinates (q, r) with cube coordinate conversions

//# Matches web/src/core/HexCoordinates.ts
[GlobalClass]
public partial class HexCoordinates : Godot.RefCounted
{
	public int Q;
	// Column (axial)
	public int R;


	// Row (axial)
	public override void _Init(int q_val = 0, int r_val = 0)
	{
		Q = q_val;
		R = r_val;
	}


	//# Cube coordinate X (derived from axial)
	public int GetX()
	{
		return Q;
	}


	//# Cube coordinate Y (derived from axial)
	public int GetY()
	{
		return  - Q - R;
	}


	//# Cube coordinate Z (derived from axial)
	public int GetZ()
	{
		return R;
	}


	//# Convert to world position at given elevation
	public Vector3 ToWorldPosition(int elevation = 0)
	{
		var x = (Q + R * 0.5) * (HexMetrics.INNER_RADIUS * 2.0);
		var z = R * (HexMetrics.OUTER_RADIUS * 1.5);
		var y = elevation * HexMetrics.ELEVATION_STEP;
		return new Vector3(x, y, z);
	}


	//# Create from world position
	public static Godot.HexCoordinates FromWorldPosition(Vector3 position)
	{
		var q_float = position.X / (HexMetrics.INNER_RADIUS * 2.0);
		var r_float = position.Z / (HexMetrics.OUTER_RADIUS * 1.5);
		q_float -= r_float * 0.5;


		// Round to nearest hex
		var q_int = Mathf.RoundToInt(q_float);
		var r_int = Mathf.RoundToInt(r_float);

		return HexCoordinates.New(q_int, r_int);
	}


	//# Calculate distance to another hex (in hex steps)
	public int DistanceTo(Godot.HexCoordinates other)
	{
		var dx = Mathf.Abs(GetX() - other.GetX());
		var dy = Mathf.Abs(GetY() - other.GetY());
		var dz = Mathf.Abs(GetZ() - other.GetZ());
		return (dx + dy + dz) / 2;
	}


	//# Get neighbor in given direction
	public Godot.HexCoordinates GetNeighbor(int direction)
	{
		var offsets = HexDirection.GetOffset(direction);
		return HexCoordinates.New(Q + offsets.X, R + offsets.Y);
	}


	//# String representation
	public override String _ToString()
	{
		return "(%d, %d)" % new Array{Q, R, };
	}


	//# Equality check
	public bool Equals(Godot.HexCoordinates other)
	{
		return Q == other.Q && R == other.R;
	}


}